pycaret/internal/tabular.py

```diff
@@ -8391,139 +8389,71 @@ def optimize_threshold(
     """
 
-    logger.info("Defining loss function")
-
-    def calculate_loss(
-        actual,
-        predicted,
-        tp_cost=true_positive,
-        tn_cost=true_negative,
-        fp_cost=false_positive,
-        fn_cost=false_negative,
-    ):
-
-        # true positives
-        tp = predicted + actual
-        tp = np.where(tp == 2, 1, 0)
-        tp = tp.sum()
+    logger.info("defining variables")
+    # get estimator name
+    model_name = _get_model_name(estimator)
 
-        # true negative
-        tn = predicted + actual
-        tn = np.where(tn == 0, 1, 0)
-        tn = tn.sum()
+    # defines grid
+    grid = np.arange(0, 1.0001, grid_interval)
 
-        # false positive
-        fp = (predicted > actual).astype(int)
-        fp = np.where(fp == 1, 1, 0)
-        fp = fp.sum()
-
-        # false negative
-        fn = (predicted < actual).astype(int)
-        fn = np.where(fn == 1, 1, 0)
-        fn = fn.sum()
-
-        total_cost = (tp_cost * tp) + (tn_cost * tn) + (fp_cost * fp) + (fn_cost * fn)
-
-        return total_cost
-
-    """
-    internal function to calculate loss ends here
-    """
-
-    grid = np.arange(0, 1, grid_interval)
+    # defines empty list
+    models_by_threshold = []
+    results_df = []
 
+    logger.info("starting optimization loop")
     # loop starts here
-
-    cost = []
-    # global optimize_results
-
-    logger.info("Iteration starts at 0")
-
     for i in grid:
-
-        pred_prob = (predicted >= i).astype(int)
-        cost.append(calculate_loss(actual, pred_prob))
-
-    optimize_results = pd.DataFrame(
-        {"Probability Threshold": grid, "Cost Function": cost}
-    )
-    fig = px.line(
-        optimize_results,
-        x="Probability Threshold",
-        y="Cost Function",
-        line_shape="linear",
-    )
-    fig.update_layout(plot_bgcolor="rgb(245,245,245)")
-    title = f"{model_name} Probability Threshold Optimization"
-
-    # calculate vertical line
-    y0 = optimize_results["Cost Function"].min()
-    y1 = optimize_results["Cost Function"].max()
-    x0 = optimize_results.sort_values(by="Cost Function", ascending=False).iloc[0][0]
-    x1 = x0
-
-    t = x0
-    if html_param:
-
-        fig.add_shape(
-            dict(
-                type="line", x0=x0, y0=y0, x1=x1, y1=y1, line=dict(color="red", width=2)
-            )
-        )
-        fig.update_layout(
-            title={
-                "text": title,
-                "y": 0.95,
-                "x": 0.45,
-                "xanchor": "center",
-                "yanchor": "top",
-            }
-        )
-        logger.info("Figure ready for render")
-        fig.show()
-    print(f"Optimized Probability Threshold: {t} | Optimized Cost Function: {y1}")
-    logger.info(
-        "optimize_threshold() succesfully completed......................................"
-    )
-
-    return float(t)
+        model = create_model_supervised(estimator, verbose=False, system=False, probability_threshold=i)
```

Comment:
this is pretty heavy. in order to optimize prediction threshold we should only need to fit once (per fold). that being said it should be refactored in a future PR

Reply:
Ok.